src/any.rs

```diff
@@ -144,6 +151,84 @@ pub enum Union {
     Map(Box<Map>),
     FnPtr(Box<FnPtr>),
     Variant(Box<Box<dyn Variant>>),
+    Shared(Box<SharedCell>),
+}
+
+/// Internal Shared Dynamic representation.
+///
+/// Created with `Dynamic::into_shared()`.
+#[derive(Clone)]
+pub struct SharedCell {
+    value_type_id: TypeId,
+    value_type_name: &'static str,
+    container: SharedMut<Dynamic>
```

Comment:
Suggest `SharedMut<dyn Any>`.  No need for this to be `Dynamic` here as the type of the `dyn Any` is already known.

Reply:
@schungx Well, there are a lot of places in the code that expecting just normal non-shared `Dynamic` rather than `Any`. In this case it is easier to do `shared_d.read_lock::<Dynamic>()` that will simply return what is inside the `SharedCell.container`. In other words it's easier to implement interoperability between normal and Shared Dynamics

Reply:
@schungx But you are right that it looks like an extra overhead. I will try to simplify it to `dyn Any` or `dyn Variant` in the end of the work. Maybe this is not that hard.

Reply:
I'd suggest you leave everything that expects a `Dynamic` to be that `Dynamic`, and only unwraps it to a `&T` at the very end when it is ready to pass to a function.

This way, you can limit the changes to mostly the `Dynamic` type itself.

Reply:
If you really want to have Rhai automatically "share" a variable simply via its access, we can compile the closure statement into multiple virtual statements:

```rust
let x = 42;
let y = 123;

let f = || x + y;

// compiles to...

let f = {
    make_shared(x);     // some operation to make 'x' shared
    make_shared(y);     // make 'y' shared
    Fn("anon$xxxxxx");
};
```

Reply:
@schungx Well, yes, this is more or less is my current plan. I want to introduce the `shared` keyword that will essential do what you shown in your first snippet. So it will always clone the parent's variables into fn's scope(through curry-ing), but if the variable appear to be already of a Shared Dynamic value, in the child scope it will work with it's inner shared value, dereferencing RefCell or RwLock at the very end. This is exactly what I'm trying to implement now.

The question is how to implement this dereferencing properly. There are a number of cases. Let me show a couple of them:

1. Registering native functions: https://github.com/schungx/rhai/blob/master/src/fn_register.rs#L104

    We obviously can't do simple `downcast_mut` here as the `data` might be ordinary Dynamic as well as Shared Dynamic. To unify both cases I used `data.write_lock` instead returning `DynamicWriteLock`: https://github.com/schungx/rhai/pull/20/files#diff-0e6bc40ce5194e38b989030b55e4946bR104 . The `DynamicWriteLock` internally holds either an ordinary reference to the ordinary `Dynamic` if the Dynamic was not shared, or it holds a `Ref`/`RwLockWriteGuard` of Shared Dynamic. For ordinary Dynamic the reference is obtained simply by applying `downcast_mut` private method under the hood. And in case of Shared Dynamic `downcast_mut` is applied too to the inner **Dynamic** value of the `RefCell`/`RwLock`, but in later stage when the `DerefMut::deref_mut` is called on top of `DynamicWriteLock`. In both cases the access to underlying `Dynamic` becomes transparent: https://github.com/schungx/rhai/pull/20/files#diff-0e6bc40ce5194e38b989030b55e4946bR140

   So, to sum up having `RefCell<Dynamic>`/`RwLock<Dynamic>` would be handy here instead of `RefCell<dyn Any>`/`RwLock<dyn Any>`, because in the case I can still use `downcast_ref/mut` under the hood.

2. Another case is dealing with Dynamic's internal data in the script. Let's assume the user wrote something like this:

   ```
   let x = shared([1, 2, 3]);

   || x[1] = 200;
   ```

   In this case we have to enrich this match-case https://github.com/schungx/rhai/blob/master/src/engine.rs#L1073 to be able to manage both cases: `Dynamic(Union::Array(...))` and `Dynamic(Union::Shared(...?))`. The second case(Shared) cannot be simply dereferenced, because `RefCell`/`RwLock` cannot be statically dereferenced. In other words it means that the `Target` needs to be extended to also be able to hold not only `Target::Ref`, but also `Target::LockGuard(DynamicWriteLock(...))`. And if the `DynamicWriteLock` would return underlying `Dynamic` itself instead of `dyn Any` we will not need to change the `Engine::get_indexed_mut`. The `get_indexed_mut` would work just the way it works currently, again, transparently dealing with Shared Dynamic.

Reply:
@schungx Regarding your second snippet, this is the goal of my next PR. When I finish and stabilize `shared` keyword in this PR and all of it's internal stuff(which may take some time) and you accept my PR :) we can work on extending of the Parser to automatically and implicitly turn all captured variables into shared:

```
let x = 42;
let y = 123;

let f = || x + y;
```

will implicitly turn into

```
let x = shared(42);
let y = shared(123);

let f = || x + y;
```

(again, when the `no_capture` is not enabled)

Reply:
> So, to sum up having RefCell<Dynamic>/RwLock<Dynamic> would be handy here instead of RefCell<dyn Any>/RwLock<dyn Any>, because in the case I can still use downcast_ref/mut under the hood.

You don't need to.  `Dynamic::downcast_ref/mut` is implemented on top of `<dyn Any>::downcast_ref/mut`.  Therefore they are the same thing.  So you can implement `read_lock` and `write_lock` to access `T` instead of `Dynamic` by using `<dyn Any>::downcast_ref/mut`.

`Dynamic` internally uses `<dyn Any>::downcast_ref/mut`, for example: https://github.com/schungx/rhai/blob/master/src/any.rs#L539

Reply:
> Target needs to be extended to also be able to hold not only Target::Ref, but also Target::LockGuard(DynamicWriteLock(...)). 

Not really.  Everything else in the Engine works with `Dynamic`, so there is no special reason why `Target` should even know about another `Dynamic` variant...  I believe there is something you're overlooking here.

But of course, you need to consider the new variant in `get_indexed_mut` and detect whether it is an array wrapped up in a lock guard.  But then it is a simple matter of just going straight to the core (like the other `Union` matches):

```rust
Dynamic(Union::Shared(shared)) => {
    let guard = shared.lock();
    if guard.type_id() == TypeId::of::<Array>() { 
        ....
    } else if if guard.type_id() == TypeId::of::<Map>() {
        ....
    } ...
}
```

Reply:
@schungx If I understand the purpose of the `Target` object correctly, it may return a reference to something that could be later mutated in expression:
```
let x = shared(#{a: 10});
x.a = 20;
```
So, once the `Engine` encounters an lhs expression `x.a` it should be able to obtain a mutable reference to `x` in order to mutate the structure when the `rhs` computed. The problem here is that if `x` is `RwLock` or `RefCell` we can't leak a mutable reference to it's guard without holding the guard object itself in the `Target`.

Reply:
Moreover, if `RwLock`/`RefCell` would hold `Any` what are we going to put inside `Target`? `Any`? Then the Engine will have to manage all possible cases of `Variant`s too in addition to `Target::Ref(&mut Dynamic)` and `Target::Value(Dynamic)`.

Reply:
Hhhmmm.... I see.  `get_indexed_mut` will return a `Dynamic`.  You're right, it probably makes more sense to have the lock hold another `Dynamic`.

Reply:
@schungx Yes, I already did: https://github.com/schungx/rhai/pull/20/files#diff-8fafaf9ae6308a3c91fbf941e818c0bcR128 :)